### PR TITLE
fix: slice backward preprocess predicates per row

### LINKED ISSUE #2852: FA4 SM103 backward preprocess predicate shape mismatch for padded head dimensions
### Summary

On B300/SM103, FA4 beta27's backward preprocess reuses a predicate covering all per-thread row modes after slicing the O/dO source and destination tensors to a single row. CuTe rejects the resulting copy because the predicate has shape `(1,(2,3))` while the data expects `(1,(3))`.

The same `flash_bwd_preprocess.py` code is present in beta29 and current `main` (`4fd8d758c1fb7d93ef2e9a316ba4db1ab7fc05fe` as fetched on 2026-09-03).

### Environment

- NVIDIA B300 / SM103
- CUDA 13.2 base image
- PyTorch 2.10.0a0+3a035ad5.nv25.12
- Python 3.13
- `flash-attn-4==4.0.0b27`
- `nvidia-cutlass-dsl==4.6.2`
- BF16, varlen causal attention
- `CUTE_DSL_SHOW_STACKTRACE=1`

### Sanitized reproducer

```python
import torch
from flash_attn.cute import flash_attn_varlen_func

torch.manual_seed(42)
q = torch.randn(256, 16, 80, device="cuda", dtype=torch.bfloat16, requires_grad=True)
k = torch.randn(256, 4, 80, device="cuda", dtype=torch.bfloat16, requires_grad=True)
v = torch.randn(256, 4, 80, device="cuda", dtype=torch.bfloat16, requires_grad=True)
cu = torch.tensor([0, 256], device="cuda", dtype=torch.int32)
result = flash_attn_varlen_func(
    q,
    k,
    v,
    cu_seqlens_q=cu,
    cu_seqlens_k=cu,
    max_seqlen_q=256,
    max_seqlen_k=256,
    causal=True,
)
out = result[0] if isinstance(result, tuple) else result
out.float().square().mean().backward()
```

Backward-preprocess compile key:

```text
dtype=BFloat16, head_dim=80, head_dim_v=80, m_block_size=128,
has_cuseqlens_q=True, has_seqused_q=False, has_dlse=False,
has_dq_accum=True, has_scale_p=False, use_padded_offsets=True,
nheads_major=False, pack_gqa=False, qhead_per_kvhead=1,
nheads_kv=1, has_cu_total_m_blocks=False
```

Compiler diagnostic:

```text
error: unknown: 'cute.copy' op expects pred to have compatible shape with: (1,(3)) but got actual predShape: (1,(2,3))
note: unknown: see current operation: "cute.copy"(...) :
  (!cute_nvgpu.atom.universal_copy<bf16, 128 b>,
   !cute.memref<bf16, gmem, align<16>, "((8,1),(3)):((1,0),(32))">,
   !cute.memref<bf16, rmem, align<16>, "((8,1),(3)):((1,0),(8))">,
   !cute.memref<i8, rmem, align<32>, "(1,(2,3)):(3,(0,1))">) -> ()
```

### Source mapping and proposed correction

`flash_attn/cute/flash_bwd_preprocess.py` constructs `tOpO` from unsliced `tOcO`, then captures it in a `partial`. Inside the `m` loop it slices both data tensors with `[None, m, None]` but does not slice the predicate.

Slicing the predicate over the same row mode makes the preprocess compile and run:

```diff
-copy = partial(copy_utils.copy, pred=tOpO)
 ...
-copy(tOgO[None, m, None], tOrO[None, m, None])
-copy(tOgdO[None, m, None], tOrdO[None, m, None])
+row_predicate = tOpO[None, m, None] if const_expr(tOpO is not None) else None
+copy_utils.copy(tOgO[None, m, None], tOrO[None, m, None], pred=row_predicate)
+copy_utils.copy(tOgdO[None, m, None], tOrdO[None, m, None], pred=row_predicate)
```

This report is limited to the backward-preprocess predicate mismatch; the reproducer may encounter other constraints later in the general SM100 backward kernel once preprocess is repaired.

### PR DESCRIPTION
**Pull request by LOGO127**
## Summary

Fixes #2852.

The backward preprocess kernel slices `O` and `dO` to one row before each
copy, but reused the unsliced head-dimension predicate. For padded head
dimensions this gives CuTe a predicate with an incompatible shape. The fix
slices the predicate with the same row index before both copies and preserves
the unpredicated path when no head-dimension padding is needed.

## Validation

- `ruff check flash_attn/cute/flash_bwd_preprocess.py`
- `ruff format --check flash_attn/cute/flash_bwd_preprocess.py`
- `py -3 -m py_compile flash_attn/cute/flash_bwd_preprocess.py`
- `git diff --check`

The FA4 GPU/CuTe test suite was not run locally because this Windows machine
does not provide the required CUDA/CuTe runtime; upstream CI should exercise
the SM103 specialization described in the issue.

AI assistance was used for drafting and local validation; I reviewed the
implementation and take responsibility for the submitted changes.

### CODE REVIEW

**Comment by LOGO127** [comment]
Thanks for the upstream fix in #2698. I checked the merged commit `94345b456854a91582f6a4b26a6feda2c8e09419` and current main `9d61d35`: `flash_bwd_preprocess.py` now slices the predicate with `[None, m, None]` before both row-wise O/dO copies, retaining `None` for the unpredicated case. That covers the complete production change proposed here, while #2698 also addresses the broader unaligned-head-dimension backward path.

I am closing this PR as superseded rather than rebasing a duplicate change. This is source-level verification, not a new local B300/GPU run or a claim that this PR was merged. I have not closed #2852, since its reporter may still want to verify the original reproducer on the updated version. Thanks to the authors and maintainers for landing the broader fix. This follow-up was prepared with AI assistance.
